File: z_lc/3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        inds = {}
        res = 0
        i = 0
        for j,c in enumerate(s):
            if c in inds:
                i = max(i, inds[c] + 1)
            inds[c] = j
            res = max(res, j - i + 1)
        return res

File: z_lc/3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.rs
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        use std::cmp;
        use std::collections::HashMap;
        
        let mut inds = HashMap::new();
        let mut res = 0;
        let mut i = 0;
        for (j,c) in s.chars().enumerate() {
            if let Some(k) = inds.get(&c) {
                i = cmp::max(i, *k + 1);
            }
            inds.insert(c, j);
            res = cmp::max(res, j - i + 1);
        }
        res as i32
    }
}